src/lib/tls/msg_client_hello.cpp

```diff
@@ -781,6 +975,109 @@ void Client_Hello_13::calculate_psk_binders(Transcript_Hash_State ths)
+                   {
+                   return scheme.is_available() &&
+                          scheme.is_compatible_with(Protocol_Version::TLS_V13);
+                   });
+
+   if(compatible_signature_schemes.empty())
+      {
+      throw TLS_Exception(Alert::HANDSHAKE_FAILURE,
+                          "Failed to agree on a signature algorithm");
+      }
+
+   // RFC 8446 4.2.3
+   //    Each SignatureScheme value lists a single signature algorithm that
+   //    the client is willing to verify.  The values are indicated in
+   //    descending order of preference.
+   for (const auto& scheme : compatible_signature_schemes)
+      {
+      // TODO: To fully support "signature_algorithm_cert", this would need to
+      //       receive two algorithm names. One for the signature algorithm used
+      //       to sign the certificate and one for the public key contained in
+      //       the certificate. Both must be supported to comply with the
+      //       client's asymmetric key requirements.
+      // Note: This requires a change in the public API of CredentialsManager.
+      //
+      // see: https://github.com/randombit/botan/issues/2714#issuecomment-1057175631
+      // see: RFC 8446 4.4.2.2
+      auto chain = creds.cert_chain_single_type(scheme.algorithm_name(), "tls-server", hostname);
+      if(!chain.empty())
+         { return chain; }
+      }
+
+   // RFC 8446 4.4.2.2
+   //    If the server cannot produce a certificate chain that is signed only
+   //    via the indicated supported algorithms, then it SHOULD continue the
+   //    handshake by sending the client a certificate chain of its choice
+   //    that may include algorithms that are not known to be supported by the
+   //    client.
+   //
+   // We rely on the application to hand out a valid (but maybe incompatible)
+   // server certificate chain. Otherwise, we have to abort.
+   throw TLS_Exception(Alert::INTERNAL_ERROR,
+                       "Application did not a provide a server certificate chain");
+   }
+
+
+std::optional<Protocol_Version> Client_Hello_13::highest_supported_version(const Policy& policy) const
+   {
+   // RFC 8446 4.2.1
+   //    The "supported_versions" extension is used by the client to indicate
+   //    which versions of TLS it supports and by the server to indicate which
+   //    version it is using. The extension contains a list of supported
+   //    versions in preference order, with the most preferred version first.
+   const auto supvers = m_data->extensions.get<Supported_Versions>();
+   BOTAN_ASSERT_NONNULL(supvers);
+
+   std::optional<Protocol_Version> result;
+
+   for(const auto& v : supvers->versions())
+      {
+      // RFC 8446 4.2.1
+      //    Servers MUST only select a version of TLS present in that extension
+      //    and MUST ignore any unknown versions that are present in that
+      //    extension.
+      if(!v.known_version() || !policy.acceptable_protocol_version(v))
+         { continue; }
```

Comment:
Should we ignore (or reject?) DTLS versions being mixed in here? Or does that happen elsewhere already.

Perhaps for RFC 8446 purposes DTLS counts as an "unknown version" of TLS but that doesn't really seem correct to me.

Reply:
DTLS as such is not explicitly filtered out anywhere. TLS 1.2 / 1.3 distinctions are typically done using `Protocol_Version::is_pre_tls_13()` or `::is_tls_13_or_later()` which transparently takes DTLS into account.

That said, DTLS 1.3 won't pass the default implementation of `Policy::acceptable_protocol_version()` and won't make it into the result vector. For reference:

```C++
bool Policy::acceptable_protocol_version(Protocol_Version version) const
   {
#if defined(BOTAN_HAS_TLS_13)
   if(version == Protocol_Version::TLS_V13 && allow_tls13())
      return true;
#endif

#if defined(BOTAN_HAS_TLS_12)
   if(version == Protocol_Version::TLS_V12 && allow_tls12())
      return true;

   if(version == Protocol_Version::DTLS_V12 && allow_dtls12())
      return true;
#endif

   return false;
   }
```